expression/builtin_compare.go

```diff
@@ -382,38 +383,68 @@ func temporalWithDateAsNumEvalType(argTp *types.FieldType) (argEvalType types.Ev
 	return
 }
 
-// GetCmpTp4MinMax gets compare type for GREATEST and LEAST and BETWEEN
-func GetCmpTp4MinMax(args []Expression) (argTp types.EvalType) {
-	datetimeFound, isAllStr := false, true
-	cmpEvalType, isStr, isTemporalWithDate := temporalWithDateAsNumEvalType(args[0].GetType())
-	if !isStr {
-		isAllStr = false
+func aggregateType(args []Expression) *types.FieldType {
+	fieldTypes := make([]*types.FieldType, len(args))
+	for i := range fieldTypes {
+		fieldTypes[i] = args[i].GetType()
 	}
-	if isTemporalWithDate {
-		datetimeFound = true
+	return types.AggFieldType(fieldTypes)
+}
+
+// ResolveType4Between resolves eval type for between expression.
+func ResolveType4Between(args [3]Expression) types.EvalType {
+	cmpTp := args[0].GetType().EvalType()
+	for i := 1; i < 3; i++ {
+		cmpTp = getBaseCmpType(cmpTp, args[i].GetType().EvalType(), nil, nil)
 	}
-	lft := args[0].GetType()
-	for i := range args {
-		rft := args[i].GetType()
-		var tp types.EvalType
-		tp, isStr, isTemporalWithDate = temporalWithDateAsNumEvalType(rft)
-		if isTemporalWithDate {
-			datetimeFound = true
+
+	hasTemporal := false
+	if cmpTp == types.ETString {
+		for _, arg := range args {
+			if types.IsTypeTemporal(arg.GetType().Tp) {
+				hasTemporal = true
+				break
+			}
 		}
-		if !isStr {
-			isAllStr = false
+		if hasTemporal {
+			cmpTp = types.ETDatetime
 		}
-		cmpEvalType = getBaseCmpType(cmpEvalType, tp, lft, rft)
-		lft = rft
 	}
-	argTp = cmpEvalType
-	if cmpEvalType.IsStringKind() {
-		argTp = types.ETString
+
+	return cmpTp
+}
+
+// resolveType4Extremum gets compare type for GREATEST and LEAST and BETWEEN (mainly for datetime).
+func resolveType4Extremum(args []Expression) types.EvalType {
+	aggType := aggregateType(args)
+
+	var temporalItem *types.FieldType
+	aggType.EvalType()
+	if types.ResultMergeType(aggType.Tp) == types.ETString {
+		for i := range args {
+			item := args[i].GetType()
+			if types.IsTemporalWithDate(item.Tp) {
+				temporalItem = item
+			}
+		}
+
+		if !types.IsTemporalWithDate(aggType.Tp) && temporalItem != nil {
+			aggType.Tp = temporalItem.Tp
+		}
+		// TODO: String charset, collation checking are needed.
 	}
-	if isAllStr && datetimeFound {
-		argTp = types.ETDatetime
+	return aggType.EvalType()
+}
+
+// unsupportedJSONComparison reports warnings while there is a JSON type in least/greatest function's arguments
+func unsupportedJSONComparison(ctx sessionctx.Context, args []Expression) {
+	for _, arg := range args {
+		tp := arg.GetType().Tp
+		if types.ResultMergeType(tp) == types.ETString && tp == mysql.TypeJSON {
```

Comment:
`types.ResultMergeType(tp) == types.ETString` is useless?

Reply:
Yes. In MySQL, there is a variant of result types, named `ROW_RESULT` which represents a row instead of a type of constant.

Reply:
Done.